File: EnterpriseRag.Api/Controllers/IngestionController.cs
using EnterpriseRag.Api.Data;
using EnterpriseRag.Api.Models;
using EnterpriseRag.Api.Services;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace EnterpriseRag.Api.Controllers
{
    [ApiController]
    [Route("api/v1/ingestion")]
    public class IngestionController : ControllerBase
    {
        private readonly IDocumentParserService _parserService;
        private readonly IChunkingService _chunkingService;
        private readonly IEmbeddingService _embeddingService;
        private readonly MongoDbContext _dbContext;
        private readonly ILogger<IngestionController> _logger;

        public IngestionController(
            IDocumentParserService parserService,
            IChunkingService chunkingService,
            IEmbeddingService embeddingService,
            MongoDbContext dbContext,
            ILogger<IngestionController> logger)
        {
            _parserService = parserService;
            _chunkingService = chunkingService;
            _embeddingService = embeddingService;
            _dbContext = dbContext;
            _logger = logger;
        }

        [HttpPost("debug-chunk")]
        public async Task<IActionResult> DebugChunk(IFormFile file)
        {
            try
            {
                if (file == null || file.Length == 0)
                {
                    return BadRequest(new { error = "No file was uploaded or the file is empty." });
                }

                if (!Path.GetExtension(file.FileName).Equals(".pdf", StringComparison.OrdinalIgnoreCase))
                {
                    return BadRequest(new { error = "Only PDF files are supported." });
                }

                _logger.LogInformation("Processing file: {FileName}, Size: {FileSize} bytes", file.FileName, file.Length);

                // Open the stream directly from the IFormFile
                using var stream = file.OpenReadStream();
                
                // Phase 1: Extract Text
                var rawText = await _parserService.ExtractTextAsync(stream);

                if (string.IsNullOrWhiteSpace(rawText))
                {
                    _logger.LogWarning("No extractable text was found in the provided PDF.");
                    return Ok(new { message = "No extractable text found in the PDF.", chunks = Array.Empty<TextChunkDto>() });
                }

                // Phase 2: Chunk the Text
                var stringChunks = _chunkingService.GenerateChunks(rawText).ToList();

                // Map the output to structured DTOs
                var structuredChunks = stringChunks.Select((text, index) => new TextChunkDto
                {
                    SequenceIndex = index,
                    Text = text,
                    // Calculate token count based on the same delimiter logic
                    TokenCount = text.Split(new[] { ' ', '\r', '\n', '\t' }, StringSplitOptions.RemoveEmptyEntries).Length
                }).ToList();

                _logger.LogInformation("Successfully processed PDF into {ChunkCount} chunks.", structuredChunks.Count);

                return Ok(new
                {
                    totalChunks = structuredChunks.Count,
                    chunks = structuredChunks
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to process PDF ingestion for chunking.");
                
                // Return a structured 500 error response
                return StatusCode(500, new 
                { 
                    error = "An unexpected error occurred while processing the file.", 
                    details = ex.Message 
                });
            }
        }

        [HttpPost("upload")]
        public async Task<IActionResult> Upload(IFormFile file)
        {
            try
            {
                if (file == null || file.Length == 0)
                {
                    return BadRequest(new { error = "No file was uploaded or the file is empty." });
                }

                if (!Path.GetExtension(file.FileName).Equals(".pdf", StringComparison.OrdinalIgnoreCase))
                {
                    return BadRequest(new { error = "Only PDF files are supported." });
                }

                _logger.LogInformation("Starting pipeline upload for file: {FileName}, Size: {FileSize} bytes", file.FileName, file.Length);

                // 1. Extract Text
                using var stream = file.OpenReadStream();
                var rawText = await _parserService.ExtractTextAsync(stream);

                if (string.IsNullOrWhiteSpace(rawText))
                {
                    _logger.LogWarning("No extractable text was found in the provided PDF: {FileName}", file.FileName);
                    return BadRequest(new { error = "No extractable text found in the PDF." });
                }

                // 2. Chunk the Text
                var stringChunks = _chunkingService.GenerateChunks(rawText).ToList();
                if (stringChunks.Count == 0)
                {
                    _logger.LogWarning("Chunking service generated 0 chunks for: {FileName}", file.FileName);
                    return BadRequest(new { error = "Failed to split document text into chunks." });
                }

                // 3. Generate Embeddings for each chunk asynchronously
                var documentChunks = new List<DocumentChunk>();
                for (int i = 0; i < stringChunks.Count; i++)
                {
                    var textChunk = stringChunks[i];
                    var embedding = await _embeddingService.GenerateEmbeddingAsync(textChunk);

                    var docChunk = new DocumentChunk
                    {
                        Id = MongoDB.Bson.ObjectId.GenerateNewId(),
                        SourceFile = file.FileName,
                        Content = textChunk,
                        SequenceIndex = i,
                        Embedding = embedding,
                        CreatedAt = DateTime.UtcNow
                    };
                    documentChunks.Add(docChunk);
                }

                // 4. Perform a bulk write operation (InsertManyAsync) to save all records to MongoDB
                _logger.LogInformation("Bulk inserting {Count} chunks into MongoDB for file {FileName}", documentChunks.Count, file.FileName);
                await _dbContext.DocumentChunks.InsertManyAsync(documentChunks);

                _logger.LogInformation("Successfully processed and persisted {FileName}", file.FileName);

                return Ok(new
                {
                    fileName = file.FileName,
                    totalChunksProcessed = documentChunks.Count,
                    status = "Successfully Persisted"
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to process PDF ingestion and storage pipeline.");
                return StatusCode(500, new
                {
                    error = "An unexpected error occurred during pipeline processing.",
                    details = ex.Message
                });
            }
        }
    }
}

File: enterprise-rag-node-api/src/controllers/ingestionController.ts
import { Request, Response } from 'express';
import { extractTextFromPdf } from '../services/pdfService';
import { generateChunks } from '../services/chunkingService';
import { generateEmbedding } from '../services/embeddingService';
import { getDocumentChunksCollection } from '../config/db';

export const uploadFile = async (req: Request, res: Response): Promise<void> => {
  try {
    const file = req.file;
    if (!file) {
      res.status(400).json({ error: "No file was uploaded." });
      return;
    }

    if (file.mimetype !== 'application/pdf') {
      res.status(400).json({ error: "Only PDF files are supported." });
      return;
    }

    const fileName = file.originalname;
    console.log(`Starting pipeline upload for file: ${fileName}`);

    const rawText = await extractTextFromPdf(file.buffer);
    if (!rawText.trim()) {
      res.status(400).json({ error: "No extractable text found in the PDF." });
      return;
    }

    const chunks = generateChunks(rawText);
    if (chunks.length === 0) {
      res.status(400).json({ error: "Failed to split document text into chunks." });
      return;
    }

    const documentChunks = [];
    for (let i = 0; i < chunks.length; i++) {
      const textChunk = chunks[i];
      const embedding = await generateEmbedding(textChunk);
      
      documentChunks.push({
        SourceFile: fileName,
        Content: textChunk,
        SequenceIndex: i,
        Embedding: embedding,
        CreatedAt: new Date()
      });
    }

    const collection = getDocumentChunksCollection();
    
    console.log(`Removing existing chunks for ${fileName}`);
    await collection.deleteMany({ SourceFile: fileName });

    console.log(`Bulk inserting ${documentChunks.length} chunks`);
    await collection.insertMany(documentChunks);

    res.status(200).json({
      fileName,
      totalChunksProcessed: documentChunks.length,
      status: "Successfully Persisted"
    });
  } catch (error: any) {
    console.error("Ingestion error:", error);
    res.status(500).json({ error: "An unexpected error occurred during pipeline processing.", details: error.message });
  }
};
